lib/OpenQA/Settings.pm

```diff
@@ -0,0 +1,55 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+package OpenQA::Settings;
+
+use strict;
+use warnings;
+
+# replace %NAME% with $settings{NAME}
+sub expand_placeholders {
+    my ($settings) = @_;
+
+    my $error_string;
+
+    for my $value (values %$settings) {
+        next unless defined $value;
+
+        # loop the value until there is no %KEY% in it
+        while ($value =~ /%(\w+)%/) {
+            my $key = $1;
+            last unless defined $settings->{$key};
+
+            my $replaced_value = $settings->{$key};
+
+# used to record already replaced value. If the value has been in this array which means there is a circular reference, then return an error message.
+            my @replaced_array = ();
+
+            while ($replaced_value =~ /%(\w+)%/) {
+                last unless defined $settings->{$1};
+                if (grep { $_ eq $1 } @replaced_array) {
+                    return "The key $1 is a circular reference, value is " . $settings->{$1};
```

Comment:
I think `"The key $1 contains a circular reference, its value is $settings->{$1}."` would sound better. And I guess `$1` and `$settings->{$1}` deserve a better variable name.